app/models/manageiq/providers/embedded_ansible/automation_manager/configuration_script_source.rb

```diff
@@ -87,8 +87,18 @@ def sync_queue(auth_user = nil)
   end
 
   def playbooks_in_git_repository
-    git_repository.update_repo
-    git_repository.entries(scm_branch, "").grep(/\.ya?ml$/)
+    [].tap do |playbooks|
+      git_repository.update_repo
+      git_repository.with_worktree do |worktree|
+        worktree.ref = scm_branch
+        worktree.blob_list do |root, entry|
+          playbook_filepath = Pathname.new("#{root}#{entry[:name]}")
```

Comment:
Minor, but it feels like the worktree should not expose rugged internal structure.  That is, perhaps it's better if .blob_list yielded root, relative_path, and content so the caller doesn't have to know to call `[:name]`

Reply:
Well, I was trying to avoid reading every file unless we absolutely needed to, hence why I am working with `worktree` to read the file when needed.

Note how I have `not_playbook_dir?` first and will eject early if true, which comes before any actually file reading in `is_a_playbook?`.

Reply:
continuing from my [comment below](https://github.com/ManageIQ/manageiq/pull/19089#discussion_r309894402), I think this can be done without introducing walk_blobs as:

```ruby
playbooks = worktree.file_list.select do |f|
  !playbook_dir?(f) && !playbook?(f, worktree)
end

...

def playbook?(file, worktree)
  return false unless file =~ /\.ya?ml$/

  worktree.read_file(file).lines.each do |line|
    return true if line =~ VALID_PLAYBOOK_CHECK
  end

  return false
end
```

Reply:
I will test, but I don't think I can use `file_list` as you have it here, for a few reasons:

First off, as mentioned in my commit message:

```
[EmbeddedAnsible] Handle nested playbooks

With the current GitRepository#entries method, the results returned are
only the top level files.

In this patch a new method, `GitWorktree#blob_list`, was added to list
only files (no directory entires) for a given ref (or the current one).
Using this in place of `.entries` allows us to find nested playbooks
properly.
```

So by doing #yourWay™, we have to do a few extra "N+1's" in git as a result:

- With `file_list`, every directory included with the list, meaning we need a separate check add to filter those out from going through `.playbook?` and `playbook_dir?`, probably requiring that we re-fetch the entry to check if it was a `Rugged::Blob` or not
- By passing the `"entry"` around instead of re-fetching it again when we read the file, we eliminate needing to do an extra lookup (`GitWorktree#fetch_entry`) when doing `GitWorktree#read_file` as shown above.

> I'm not sure what walk_blobs is getting you if you are using read_entry anyway. That is, you can also just use file_list and then use read_file (which uses read_entry under the covers).

Here might be a easier to visualize example that you can run (or just see the output of):

```ruby
require_relative "../lib/git_worktree"
require_relative "../spec/support/fake_ansible_repo"

repo_path      = File.expand_path "./test_repo_2", __dir__
file_structure = <<~REPO
  roles/
    azure_stuffs/
      tasks/
        main.yml
      defaults/
        main.yml
    requirements.yml
  playbook.yml
REPO

Spec::Support::FakeAnsibleRepo.generate repo_path, file_structure

worktree = GitWorktree.new(:path => File.expand_path("./test_repo_2", __dir__))

puts
puts "GitWorktree#file_list"
puts "---------------------"
puts worktree.file_list
puts
puts "GitWorktree#blob_list"
puts "---------------------"
puts worktree.blob_list

# sample_out
#
# $ ruby tmp/test_fake_ansible_repo.rb
# GitWorktree#file_list
# ---------------------
# playbook.yml
# roles
# roles/azure_stuffs
# roles/azure_stuffs/defaults
# roles/azure_stuffs/defaults/main.yml
# roles/azure_stuffs/tasks
# roles/azure_stuffs/tasks/main.yml
# roles/requirements.yml
#
# GitWorktree#blob_list
# ---------------------
# playbook.yml
# roles/azure_stuffs/defaults/main.yml
# roles/azure_stuffs/tasks/main.yml
# roles/requirements.yml
```

As you can see, `roles/`, `roles/azure_stuffs`, and `roles/azure_stuffs/{defaults,tasks}` are all omitted when using `walk_blobs` under the hood, saving LOC and excessive processing in ruby.

Reply:
Went with suggestion in the https://github.com/ManageIQ/manageiq/pull/19089#discussion_r309900233 to avoid to much more back and forth with this one, so this should be addressed now.